kernel/idle.c

```diff
@@ -152,6 +157,24 @@ void idle(void *unused1, void *unused2, void *unused3)
 #endif
 
 	while (true) {
+		/* Lock interrupts to atomically check if to_abort is non-NULL,
+		 * and if so clear it
+		 */
+		int key = arch_irq_lock();
+		struct k_thread *to_abort = cpu->pending_abort;
+
+		if (to_abort) {
+			cpu->pending_abort = NULL;
+			arch_irq_unlock(key);
+
+			LOG_DBG("idle %p aborting thread %p",
+				_current, to_abort);
+
+			z_thread_single_abort(to_abort);
```

Comment:
Doing this outside the look seems racy?  Maybe not a problem in practice because of the flag state, but does z_thread_single_abort() really need to be called with interrupts unmasked?

Reply:
> Doing this outside the look seems racy?

I think this is OK. We've atomically checked and stashed the `cpu->pending_abort` value to a stack variable. If we get preempted before `z_thread_single_abort()` is called, and another thread aborts, `cpu->pending_abort` will be set and we'll get right back here. We'll then abort the first thread, and then repeat the loop iteration to get the second one. I'll add a comment explaining this.

>  does z_thread_single_abort() really need to be called with interrupts unmasked?

It doesn't. I thought it would be a little better for latency, especially if thread->fn_abort has a function attached.